File: sdk/python/src/aegis_sdk/pipeline.py
"""Defense pipeline orchestration for chat completions."""

from __future__ import annotations

import uuid
from typing import Any

import httpx

from aegis_sdk.errors import AegisApprovalRequiredError, AegisPolicyBlockedError, AegisProviderError


def _trace() -> dict[str, str]:
    return {"trace_id": str(uuid.uuid4()), "request_id": str(uuid.uuid4())}


def _user_text(messages: list[dict[str, Any]]) -> str:
    parts: list[str] = []
    for msg in messages:
        if msg.get("role") == "user" and msg.get("content"):
            parts.append(str(msg["content"]))
    if not parts:
        raise ValueError("messages must include at least one user message")
    return "\n\n".join(parts)


class DefensePipeline:
    """Runs input → policy → model-router → output → policy for chat requests."""

    def __init__(
        self,
        *,
        input_defense_url: str,
        output_defense_url: str,
        policy_engine_url: str,
        model_router_url: str,
        agent_gate_url: str,
        tenant_id: str = "default",
        timeout: float = 60.0,
    ) -> None:
        self.input_defense_url = input_defense_url.rstrip("/")
        self.output_defense_url = output_defense_url.rstrip("/")
        self.policy_engine_url = policy_engine_url.rstrip("/")
        self.model_router_url = model_router_url.rstrip("/")
        self.agent_gate_url = agent_gate_url.rstrip("/")
        self.tenant_id = tenant_id
        self.timeout = timeout

    async def chat_completions(
        self,
        *,
        model: str,
        messages: list[dict[str, Any]],
        provider: str | None = None,
        stream: bool = False,
        temperature: float | None = None,
        max_tokens: int | None = None,
        trace: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        if stream:
            raise AegisProviderError(
                "Defended streaming is not supported: output defense requires the "
                "complete assistant response before content is released to the client",
                error_type="streaming_unsupported",
            )

        trace_ctx = trace or _trace()
        user_text = _user_text(messages)

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            input_resp = await self._post(
                client,
                f"{self.input_defense_url}/analyze",
                {
                    "tenant_id": self.tenant_id,
                    "trace": trace_ctx,
                    "text": user_text,
                },
            )
            input_verdict = input_resp["verdict"]
            self._check_input_verdict(input_verdict)

            policy_in = await self._post(
                client,
                f"{self.policy_engine_url}/v1/evaluate/input",
                {
                    "tenant_id": self.tenant_id,
                    "mode": "enforce",
                    "trace": trace_ctx,
                    "input_verdict": input_verdict,
                },
            )
            self._check_policy_decision(policy_in["decision"], layer="input")

            router_body: dict[str, Any] = {
                "model": model,
                "messages": messages,
            }
            if provider:
                router_body["provider"] = provider
            if temperature is not None:
                router_body["temperature"] = temperature
            if max_tokens is not None:
                router_body["max_tokens"] = max_tokens

            llm_resp = await self._post(
                client,
                f"{self.model_router_url}/v1/chat/completions",
                router_body,
                provider_errors=True,
            )

            content = llm_resp["choices"][0]["message"]["content"]

            output_resp = await self._post(
                client,
                f"{self.output_defense_url}/analyze",
                {
                    "tenant_id": self.tenant_id,
                    "trace": trace_ctx,
                    "content": content,
                    "original_prompt": user_text,
                },
            )
            output_verdict = output_resp["verdict"]
            self._check_output_verdict(output_verdict)

            policy_out = await self._post(
                client,
                f"{self.policy_engine_url}/v1/evaluate/output",
                {
                    "tenant_id": self.tenant_id,
                    "mode": "enforce",
                    "trace": trace_ctx,
                    "output_verdict": output_verdict,
                },
            )
            self._check_policy_decision(policy_out["decision"], layer="output")

            final_content = output_verdict.get("redacted_content") or content
            llm_resp["choices"][0]["message"]["content"] = final_content
            llm_resp.setdefault("aegis", {})
            llm_resp["aegis"].update(
                {
                    "trace_id": trace_ctx["trace_id"],
                    "request_id": trace_ctx["request_id"],
                    "input_verdict": input_verdict,
                    "output_verdict": output_verdict,
                    "input_policy": policy_in["decision"],
                    "output_policy": policy_out["decision"],
                }
            )
            return llm_resp

    async def evaluate_tool(
        self,
        *,
        tool_call: dict[str, Any],
        trace: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        trace_ctx = trace or _trace()
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            resp = await self._post(
                client,
                f"{self.agent_gate_url}/v1/evaluate",
                {
                    "tenant_id": self.tenant_id,
                    "mode": "enforce",
                    "trace": trace_ctx,
                    "tool_call": tool_call,
                },
            )
            decision = resp.get("decision", {})
            status = str(decision.get("status", ""))
            if status == "AWAITING_HUMAN_APPROVAL":
                raise AegisApprovalRequiredError(
                    "Tool call requires human approval",
                    approval_id=str(decision.get("approval_request_id", "")),
                    tool_name=tool_call.get("tool_name"),
                    details=resp,
                )
            if status == "DENIED":
                raise AegisPolicyBlockedError(
                    decision.get("denial_reason") or "Tool call denied by agent-gate",
                    layer="tool_gate",
                    action=status,
                    details=resp,
                )
            resp.setdefault("aegis", {})
            resp["aegis"]["trace_id"] = trace_ctx["trace_id"]
            resp["aegis"]["request_id"] = trace_ctx["request_id"]
            return resp

    async def _post(
        self,
        client: httpx.AsyncClient,
        url: str,
        payload: dict[str, Any],
        *,
        provider_errors: bool = False,
    ) -> dict[str, Any]:
        resp = await client.post(url, json=payload)
        if resp.status_code >= 400:
            body: dict[str, Any] = {}
            try:
                body = resp.json()
            except Exception:
                body = {"error": resp.text}
            if provider_errors:
                raise self._provider_error(resp.status_code, body)
            raise httpx.HTTPStatusError(
                f"{url} returned {resp.status_code}",
                request=resp.request,
                response=resp,
            )
        return resp.json()  # type: ignore[no-any-return]

    def _provider_error(self, status_code: int, body: dict[str, Any]) -> AegisProviderError:
        aegis = body.get("aegis") or {}
        model_error = aegis.get("model_error") or {}
        return AegisProviderError(
            str(body.get("error", "upstream provider error")),
            status_code=status_code,
            provider=model_error.get("provider"),
            model=model_error.get("rejected_model"),
            error_type=model_error.get("error_type") or "provider_error",
            details=body,
        )

    @staticmethod
    def _check_input_verdict(verdict: dict[str, Any]) -> None:
        action = str(verdict.get("action", "")).upper()
        if action == "BLOCK":
            raise AegisPolicyBlockedError(
                "Input blocked by input-defense",
                layer="input_defense",
                action=action,
                fused_score=verdict.get("fused_score"),
                details={"input_verdict": verdict},
            )

    @staticmethod
    def _check_output_verdict(verdict: dict[str, Any]) -> None:
        action = str(verdict.get("action", "")).upper()
        if action == "BLOCK":
            raise AegisPolicyBlockedError(
                "Output blocked by output-defense",
                layer="output_defense",
                action=action,
                fused_score=verdict.get("fused_score"),
                details={"output_verdict": verdict},
            )

    @staticmethod
    def _check_policy_decision(decision: dict[str, Any], *, layer: str) -> None:
        action = str(decision.get("action", "")).lower()
        if action == "block":
            raise AegisPolicyBlockedError(
                decision.get("block_reason") or f"Blocked by policy ({layer})",
                layer=f"policy_{layer}",
                policy_action=action,
                details={"policy_decision": decision},
            )

File: gateway/internal/pipeline/pipeline.go
package pipeline

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/aegis-platform/aegis/gateway/internal/config"
)

// ChatRequest mirrors the OpenAI-compatible gateway chat body.
type ChatRequest struct {
	Model       string           `json:"model"`
	Messages    []map[string]any `json:"messages"`
	Provider    string           `json:"provider,omitempty"`
	Stream      bool             `json:"stream,omitempty"`
	Temperature *float64         `json:"temperature,omitempty"`
	MaxTokens   *int             `json:"max_tokens,omitempty"`
}

// ToolEvaluateRequest is the agent-gate tool evaluation body.
type ToolEvaluateRequest struct {
	ToolCall map[string]any `json:"tool_call"`
}

// Pipeline orchestrates input → policy → model-router → output → policy.
type Pipeline struct {
	cfg    config.Config
	client *http.Client
}

func New(cfg config.Config) *Pipeline {
	return &Pipeline{
		cfg: cfg,
		client: &http.Client{
			Timeout: time.Duration(cfg.HTTPTimeoutSeconds) * time.Second,
		},
	}
}

func (p *Pipeline) ChatCompletions(ctx context.Context, req ChatRequest, trace map[string]string) (map[string]any, error) {
	if req.Stream {
		return nil, &StreamingUnsupportedError{}
	}
	if req.Model == "" {
		req.Model = p.cfg.DefaultModel
	}

	userText, err := userTextFromMessages(req.Messages)
	if err != nil {
		return nil, err
	}
	if trace == nil {
		trace = newTrace()
	}

	inputResp, err := p.postJSON(ctx, p.cfg.InputDefenseURL+"/analyze", map[string]any{
		"tenant_id": p.cfg.DefaultTenantID,
		"trace":     trace,
		"text":      userText,
	})
	if err != nil {
		return nil, err
	}
	inputVerdict, _ := inputResp["verdict"].(map[string]any)
	if err := checkInputVerdict(inputVerdict); err != nil {
		return nil, err
	}

	policyIn, err := p.postJSON(ctx, p.cfg.PolicyEngineURL+"/v1/evaluate/input", map[string]any{
		"tenant_id":     p.cfg.DefaultTenantID,
		"mode":          "enforce",
		"trace":         trace,
		"input_verdict": inputVerdict,
	})
	if err != nil {
		return nil, err
	}
	if decision, ok := policyIn["decision"].(map[string]any); ok {
		if err := checkPolicyDecision(decision, "input"); err != nil {
			return nil, err
		}
	}

	routerBody := map[string]any{
		"model":    req.Model,
		"messages": req.Messages,
	}
	if req.Provider != "" {
		routerBody["provider"] = req.Provider
	}
	if req.Temperature != nil {
		routerBody["temperature"] = *req.Temperature
	}
	if req.MaxTokens != nil {
		routerBody["max_tokens"] = *req.MaxTokens
	}

	llmResp, err := p.postJSON(ctx, p.cfg.ModelRouterURL+"/v1/chat/completions", routerBody)
	if err != nil {
		if perr, ok := err.(*ProviderError); ok {
			return nil, perr
		}
		return nil, err
	}

	content, err := assistantContent(llmResp)
	if err != nil {
		return nil, err
	}

	outputResp, err := p.postJSON(ctx, p.cfg.OutputDefenseURL+"/analyze", map[string]any{
		"tenant_id":       p.cfg.DefaultTenantID,
		"trace":           trace,
		"content":         content,
		"original_prompt": userText,
	})
	if err != nil {
		return nil, err
	}
	outputVerdict, _ := outputResp["verdict"].(map[string]any)
	if err := checkOutputVerdict(outputVerdict); err != nil {
		return nil, err
	}

	policyOut, err := p.postJSON(ctx, p.cfg.PolicyEngineURL+"/v1/evaluate/output", map[string]any{
		"tenant_id":      p.cfg.DefaultTenantID,
		"mode":           "enforce",
		"trace":          trace,
		"output_verdict": outputVerdict,
	})
	if err != nil {
		return nil, err
	}
	if decision, ok := policyOut["decision"].(map[string]any); ok {
		if err := checkPolicyDecision(decision, "output"); err != nil {
			return nil, err
		}
	}

	finalContent := content
	if redacted, ok := outputVerdict["redacted_content"].(string); ok && redacted != "" {
		finalContent = redacted
	}
	setAssistantContent(llmResp, finalContent)

	aegis, _ := llmResp["aegis"].(map[string]any)
	if aegis == nil {
		aegis = map[string]any{}
		llmResp["aegis"] = aegis
	}
	aegis["trace_id"] = trace["trace_id"]
	aegis["request_id"] = trace["request_id"]
	aegis["input_verdict"] = inputVerdict
	aegis["output_verdict"] = outputVerdict
	aegis["input_policy"] = policyIn["decision"]
	aegis["output_policy"] = policyOut["decision"]

	return llmResp, nil
}

func (p *Pipeline) EvaluateTool(ctx context.Context, toolCall map[string]any, trace map[string]string) (map[string]any, error) {
	if trace == nil {
		trace = newTrace()
	}
	resp, err := p.postJSON(ctx, p.cfg.AgentGateURL+"/v1/evaluate", map[string]any{
		"tenant_id": p.cfg.DefaultTenantID,
		"mode":      "enforce",
		"trace":     trace,
		"tool_call": toolCall,
	})
	if err != nil {
		return nil, err
	}
	decision, _ := resp["decision"].(map[string]any)
	status, _ := decision["status"].(string)
	switch status {
	case "AWAITING_HUMAN_APPROVAL":
		approvalID, _ := decision["approval_request_id"].(string)
		toolName, _ := toolCall["tool_name"].(string)
		return nil, &ApprovalRequiredError{
			Message:    "Tool call requires human approval",
			ApprovalID: approvalID,
			ToolName:   toolName,
			Details:    resp,
		}
	case "DENIED":
		reason, _ := decision["denial_reason"].(string)
		if reason == "" {
			reason = "Tool call denied by agent-gate"
		}
		return nil, &PolicyBlockedError{
			Message: reason,
			Layer:   "tool_gate",
			Action:  status,
			Details: resp,
		}
	}
	aegis, _ := resp["aegis"].(map[string]any)
	if aegis == nil {
		aegis = map[string]any{}
		resp["aegis"] = aegis
	}
	aegis["trace_id"] = trace["trace_id"]
	aegis["request_id"] = trace["request_id"]
	return resp, nil
}

func (p *Pipeline) postJSON(ctx context.Context, url string, payload map[string]any) (map[string]any, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)

	if resp.StatusCode >= 400 {
		var parsed map[string]any
		_ = json.Unmarshal(raw, &parsed)
		if parsed == nil {
			parsed = map[string]any{"error": string(raw)}
		}
		if url == p.cfg.ModelRouterURL+"/v1/chat/completions" {
			return nil, providerErrorFromBody(resp.StatusCode, parsed)
		}
		errText, _ := parsed["error"].(string)
		if errText == "" {
			errText = string(raw)
		}
		return nil, &HTTPError{URL: url, StatusCode: resp.StatusCode, Body: errText}
	}

	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, fmt.Errorf("decode %s: %w", url, err)
	}
	return out, nil
}

func providerErrorFromBody(statusCode int, body map[string]any) *ProviderError {
	errMsg, _ := body["error"].(string)
	aegis, _ := body["aegis"].(map[string]any)
	modelErr, _ := aegis["model_error"].(map[string]any)
	errType := "provider_error"
	if modelErr != nil {
		if t, ok := modelErr["error_type"].(string); ok && t != "" {
			errType = t
		}
	}
	provider, _ := modelErr["provider"].(string)
	model, _ := modelErr["rejected_model"].(string)
	return &ProviderError{
		Message:    errMsg,
		StatusCode: statusCode,
		Provider:   provider,
		Model:      model,
		ErrorType:  errType,
		Details:    body,
	}
}

func newTrace() map[string]string {
	return map[string]string{
		"trace_id":   randomID(),
		"request_id": randomID(),
	}
}

func randomID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

func userTextFromMessages(messages []map[string]any) (string, error) {
	var parts []string
	for _, msg := range messages {
		role, _ := msg["role"].(string)
		content, _ := msg["content"].(string)
		if role == "user" && content != "" {
			parts = append(parts, content)
		}
	}
	if len(parts) == 0 {
		return "", fmt.Errorf("messages must include at least one user message")
	}
	return joinParts(parts), nil
}

func joinParts(parts []string) string {
	out := parts[0]
	for i := 1; i < len(parts); i++ {
		out += "\n\n" + parts[i]
	}
	return out
}

func assistantContent(llmResp map[string]any) (string, error) {
	choices, ok := llmResp["choices"].([]any)
	if !ok || len(choices) == 0 {
		return "", fmt.Errorf("model-router response missing choices")
	}
	choice, ok := choices[0].(map[string]any)
	if !ok {
		return "", fmt.Errorf("invalid choice shape")
	}
	message, ok := choice["message"].(map[string]any)
	if !ok {
		return "", fmt.Errorf("invalid message shape")
	}
	content, _ := message["content"].(string)
	return content, nil
}

func setAssistantContent(llmResp map[string]any, content string) {
	choices, ok := llmResp["choices"].([]any)
	if !ok || len(choices) == 0 {
		return
	}
	choice, ok := choices[0].(map[string]any)
	if !ok {
		return
	}
	message, ok := choice["message"].(map[string]any)
	if !ok {
		message = map[string]any{}
		choice["message"] = message
	}
	message["content"] = content
}

func checkInputVerdict(verdict map[string]any) error {
	action, _ := verdict["action"].(string)
	if action == "BLOCK" {
		score, _ := verdict["fused_score"].(float64)
		return &PolicyBlockedError{
			Message:    "Input blocked by input-defense",
			Layer:      "input_defense",
			Action:     action,
			FusedScore: score,
			Details:    map[string]any{"input_verdict": verdict},
		}
	}
	return nil
}

func checkOutputVerdict(verdict map[string]any) error {
	action, _ := verdict["action"].(string)
	if action == "BLOCK" {
		score, _ := verdict["fused_score"].(float64)
		return &PolicyBlockedError{
			Message:    "Output blocked by output-defense",
			Layer:      "output_defense",
			Action:     action,
			FusedScore: score,
			Details:    map[string]any{"output_verdict": verdict},
		}
	}
	return nil
}

func checkPolicyDecision(decision map[string]any, layer string) error {
	action, _ := decision["action"].(string)
	if action == "block" {
		reason, _ := decision["block_reason"].(string)
		if reason == "" {
			reason = fmt.Sprintf("Blocked by policy (%s)", layer)
		}
		return &PolicyBlockedError{
			Message:      reason,
			Layer:        "policy_" + layer,
			PolicyAction: action,
			Details:      map[string]any{"policy_decision": decision},
		}
	}
	return nil
}
